plugin/keyring_vault/vault_curl.cc

```diff
@@ -0,0 +1,177 @@
+                                  "Please make sure that Vault's CA certificate is trusted by the machine from "
+                                  "which you intend to connect to Vault."); 
+  }
+  return FALSE;
+}
+
+my_bool Vault_curl::reset_curl_session()
+{
+  CURLcode curl_res = CURLE_OK;
+  curl_easy_reset(curl);
+  read_data_ss.str("");
+  read_data_ss.clear();
+  curl_errbuf[0] = '\0';
+  if (list != NULL)
+  {
+    curl_slist_free_all(list);
+    list = NULL;
+  }
+
+  if ((list = curl_slist_append(list, token_header.c_str())) == NULL ||
+      (list = curl_slist_append(list, "Content-Type: application/json")) == NULL ||
+      (curl_res = curl_easy_setopt(curl, CURLOPT_ERRORBUFFER, curl_errbuf)) != CURLE_OK ||
+      (curl_res = curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, write_response_memory)) != CURLE_OK ||
+      (curl_res = curl_easy_setopt(curl, CURLOPT_WRITEDATA, reinterpret_cast<void*>(&read_data_ss))) != CURLE_OK ||
+      (curl_res = curl_easy_setopt(curl, CURLOPT_HTTPHEADER, list)) != CURLE_OK ||
+      (curl_res = curl_easy_setopt(curl, CURLOPT_SSL_VERIFYPEER, 1)) != CURLE_OK ||
+      (curl_res = curl_easy_setopt(curl, CURLOPT_SSL_VERIFYHOST, 2L)) != CURLE_OK ||
+      (vault_ca.empty() == false &&
+       (curl_res = curl_easy_setopt(curl, CURLOPT_CAINFO, vault_ca.c_str())) != CURLE_OK
+      ) ||
+      (curl_res = curl_easy_setopt(curl, CURLOPT_USE_SSL, CURLUSESSL_ALL)) != CURLE_OK)
+  {
+    logger->log(MY_ERROR_LEVEL, get_error_from_curl(curl_res).c_str());
+    return TRUE;
+  }
+  return FALSE;
+}
+
+my_bool Vault_curl::list_keys(std::string *response)
+{
+  CURLcode curl_res = CURLE_OK;
+  curl_easy_reset(curl);
+  long http_code = 0;
+
+  if (reset_curl_session() ||
+      (curl_res = curl_easy_setopt(curl, CURLOPT_URL, (vault_url + "?list=true").c_str())) != CURLE_OK ||
+      (curl_res = curl_easy_perform(curl)) != CURLE_OK ||
+      (curl_res = curl_easy_getinfo (curl, CURLINFO_RESPONSE_CODE, &http_code)) != CURLE_OK)
+  {
+    logger->log(MY_ERROR_LEVEL,
+                get_error_from_curl(curl_res).c_str());
+    return TRUE;
+  }
+  if (http_code == 404)
+  {
+    *response=""; //no keys found
+    return FALSE; 
+  }
+  *response = read_data_ss.str();
+  return http_code != 200 && http_code != 202; //200 and 202 are success return code
+}
+
+my_bool Vault_curl::write_key(IKey *key, std::string *response)
+{
+  //base64 encoding
+  uint64 memory_needed = base64_needed_encoded_length(key->get_key_data_size());
+  char *base64_encoded_key_data = new char[memory_needed];
+  if (base64_encode((const char*)key->get_key_data(), key->get_key_data_size(), base64_encoded_key_data) != 0)
+  {
+    delete[] base64_encoded_key_data;
+    logger->log(MY_ERROR_LEVEL, "Could not encode a key in base64");
+    return TRUE;
+  }
+  char* new_end = std::remove(base64_encoded_key_data, base64_encoded_key_data + memory_needed, '\n');
```

Comment:
In my opinion, this is a minor change which can be made invisible for existing code.
It will make this function more flexible.
A lot of base64 libraries have this option as configuration parameter. For instance
https://msdn.microsoft.com/ru-ru/library/8f9a8s97%28v=vs.110%29.aspx?f=255&MSPPError=-2147217396